### ISSUE TITLE
HinSAGE link existence prediction?

### ISSUE THREAD

**Issue by omrishsu** [stellargraph/stellargraph#1803]
From the [demo of HinSAGE link prediction](https://stellargraph.readthedocs.io/en/stable/demos/link-prediction/hinsage-link-prediction.html) I’ve understood that the model can predict the label of a link between two nodes.

I’m looking to know if a link should be existing (binary).

For example, let’s say I have a list of users and movies they watched and I want to find (predict) missing links (movies that a user watched). There is no link between a user that did not watch a movie. So, I’m not looking to learn the label of the link but to predict if a user saw a movie (the existence of the link).

I thought about changing the data so users that "saw" a movie will be linked as rank 5 and users that didn’t will be linked as rank 1 and then using the HinSAGE as is. However, that means adding a lot of edges. In my dataset I have 130K users and movies with 250K edges (each user has 0 to 3 movies), connection all to all will create an unreasonable number of edges.
There is a better alternative?

Moreover, I also have a user to user links in my graph that I want to keep (since a connection between 2 users can reflect on a link between a user and a movie, like a family watching movies together). However, I want to predict only the links between users and movies. 
Can i "signal" HinSAGE the types of links to predict?

**Comment by huonw** [comment]
A link prediction task is often used to predict link existence. Many of the other examples in https://stellargraph.readthedocs.io/en/stable/demos/link-prediction/index.html do this.

Instead of adding all of the didn't watch edges, the typical way to do this is to just select some for training and evaluation but keep them out of the graph. This can be done via random negative sampling.

https://web.archive.org/web/20200715071324/https://community.stellargraph.io/t/hinsage-link-prediction-and-imbalanced-classification/136 contains a lot of discussion about HinSAGE link prediction that may be relevant to you. (My comments with `train_test_split` in them may be the most useful.)

>  Can i "signal" HinSAGE the types of links to predict?

HinSAGE link prediction requires specifying the "head node types", which are the node types of the source and destination, which is "user" and "movie" in this case. This means that it'll only do predictions for edges between those types, and so only use the "user"/"user" edges for informing the model.

---

Does this help?

**Comment by Anujay-Saraf** [comment]
Hi @huonw  - Thanks for answering the same, I have one more query.. How to implement the HinSAGE for a custom dataset where I am unable to set the head node types in Graph -G , please share any working example to create the same for any different dataset other then user and movie
